File: pkgs/go/tests/mock/interface.go
// Package mock provides a mock IncludeKit engine for testing without WASM
package mock

import (
	"github.com/bold-minds/includekit-spec/go/types"
)

// AppSchema is engine-specific (not in universal format spec)
type AppSchema struct {
	Version int     `json:"version"`
	Models  []Model `json:"models"`
}

// Model represents a model in the schema
type Model struct {
	Name      string     `json:"name"`
	ID        IDConfig   `json:"id"`
	Relations []Relation `json:"relations,omitempty"`
}

// IDConfig represents ID field configuration
type IDConfig struct {
	Kind string `json:"kind"`
}

// Relation represents a model relation
type Relation struct {
	Name   string `json:"name"`
	Target string `json:"target"`
	Kind   string `json:"kind"`
}

// AddQueryRequest wraps a shape with optional result hint
type AddQueryRequest struct {
	Shape      types.Statement          `json:"shape"`
	ResultHint map[string][]interface{} `json:"result_hint,omitempty"`
}

// AddQueryResponse contains shape ID and dependencies
type AddQueryResponse struct {
	ShapeID      string             `json:"shape_id"`
	Dependencies types.Dependencies `json:"dependencies"`
}

// ShapeIDResponse contains the computed shape ID
type ShapeIDResponse struct {
	ShapeID string `json:"shape_id"`
}

// InvalidateResponse contains shape IDs to evict
type InvalidateResponse struct {
	Evict []string `json:"evict"`
}

// ExplainRequest contains mutation and shape ID for explanation
type ExplainRequest struct {
	Mutation types.Mutation `json:"mutation"`
	ShapeID  string         `json:"shape_id"`
}

// ExplainResponse explains why a shape would be invalidated
type ExplainResponse struct {
	Invalidate bool     `json:"invalidate"`
	Reasons    []string `json:"reasons"`
}

// VersionInfo contains engine version information
type VersionInfo struct {
	Core     string `json:"core"`
	Contract string `json:"contract"`
	ABI      string `json:"abi"`
}

// Engine interface matching WASM exports
type Engine interface {
	SetSchema(schema AppSchema) error
	ComputeShapeID(statement types.Statement) (ShapeIDResponse, error)
	AddQuery(request AddQueryRequest) (AddQueryResponse, error)
	Invalidate(mutation types.Mutation) (InvalidateResponse, error)
	ExplainInvalidation(request ExplainRequest) (ExplainResponse, error)
	Reset()
	GetVersion() VersionInfo
}

File: pkgs/ts/tests/src/mock/interface.ts
/**
 * Engine-specific types for IncludeKit mock engine
 * These types are NOT in the universal format schema
 */

import type {
  Statement,
  Mutation,
  Dependencies
} from '@includekit/spec';

/**
 * Application schema (engine-specific)
 */
export interface AppSchema {
  version: number;
  models: Array<{
    name: string;
    id: { kind: string };
    relations?: Array<{
      name: string;
      target: string;
      kind: string;
    }>;
  }>;
}

/**
 * Request to add a query and track its dependencies
 */
export interface AddQueryRequest {
  shape: Statement;
  result_hint?: Record<string, any[]>;
}

/**
 * Response from addQuery
 */
export interface AddQueryResponse {
  shape_id: string;
  dependencies: Dependencies;
}

/**
 * Response from computeShapeId
 */
export interface ShapeIdResponse {
  shape_id: string;
}

/**
 * Response from invalidate
 */
export interface InvalidateResponse {
  evict: string[];
}

/**
 * Request to explain invalidation
 */
export interface ExplainRequest {
  mutation: Mutation;
  shape_id: string;
}

/**
 * Response from explainInvalidation
 */
export interface ExplainResponse {
  invalidate: boolean;
  reasons: string[];
}

/**
 * Version information
 */
export interface VersionInfo {
  core: string;
  contract: string;
  abi: string;
}

/**
 * Engine interface matching WASM exports
 */
export interface IIncludeKitEngine {
  setSchema(schema: AppSchema): void;
  computeShapeId(statement: Statement): ShapeIdResponse;
  addQuery(request: AddQueryRequest): AddQueryResponse;
  invalidate(mutation: Mutation): InvalidateResponse;
  explainInvalidation(request: ExplainRequest): ExplainResponse;
  reset(): void;
  getVersion(): VersionInfo;
}
